File: Python/G4/G4.py
from pathlib import Path

def ab_o_abb(s: str) -> bool:
    return s in ("ab", "abb")

def main():
    archivo = Path("G4.txt")
    if not archivo.exists():
        print("[ADVERTENCIA] No se encontró G4.txt")
        return
    with archivo.open("r", encoding="utf-8") as f:
        for i, linea in enumerate(f, start=1):
            s = linea.strip()
            acepta = ab_o_abb(s)
            veredicto = "acepta" if acepta else "NO acepta"
            mostrar = s if s else "ε"

            print(f"[G4][{i}] {mostrar} => {veredicto}")

main()

File: C/G4/G4.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

bool ab_o_abb(const char *s) {
    return strcmp(s, "ab") == 0 || strcmp(s, "abb") == 0;
}

void chomp(char *s) {
    int n = strlen(s);
    while (n > 0 && (s[n-1] == '\n' || s[n-1] == '\r')) {
        s[n-1] = '\0';
        n--;
    }
}

int main() {
    FILE *f = fopen("G4.txt", "r");
    if (!f) { printf("[ADVERTENCIA] No se encontró G4.txt\n"); return 0; }
    char linea[256];
    int i = 0;
    while (fgets(linea, sizeof(linea), f)) {
        chomp(linea);
        i++;
        bool acepta = ab_o_abb(linea);
        printf("[G4][%d] %s => %s\n", i, linea[0] ? linea : "ε", acepta ? "acepta" : "NO acepta");
    }
    fclose(f);
    return 0;
}
